src/worker/emscripten-utils.ts

```diff
@@ -0,0 +1,97 @@
+/**
+ * @license
+ * Copyright 2010 The Emscripten Authors
+ * SPDX-License-Identifier: MIT
+ */
+
+export function intArrayFromString(stringy: string, dontAddNull: boolean, length: number): number[] {
+  var len = length > 0 ? length : lengthBytesUTF8(stringy) + 1;
+  var u8array = new Array(len);
+  var numBytesWritten = stringToUTF8Array(stringy, u8array, 0, u8array.length);
+  if (dontAddNull) u8array.length = numBytesWritten; // dontAddNull == true
+  return u8array;
+}
+
+function lengthBytesUTF8(str: string) {
+  var len = 0;
+  for (var i = 0; i < str.length; ++i) {
+    var u = str.charCodeAt(i);
+    if (u >= 55296 && u <= 57343) u = (65536 + ((u & 1023) << 10)) | (str.charCodeAt(++i) & 1023);
+    if (u <= 127) ++len;
+    else if (u <= 2047) len += 2;
+    else if (u <= 65535) len += 3;
+    else len += 4;
+  }
+  return len;
+}
+
+function stringToUTF8Array(str: string, heap: number[], outIdx: number, maxBytesToWrite: number) {
+  if (!(maxBytesToWrite > 0)) return 0;
+  var startIdx = outIdx;
+  var endIdx = outIdx + maxBytesToWrite - 1;
+  for (var i = 0; i < str.length; ++i) {
+    var u = str.charCodeAt(i);
+    if (u >= 55296 && u <= 57343) {
+      var u1 = str.charCodeAt(++i);
+      u = (65536 + ((u & 1023) << 10)) | (u1 & 1023);
+    }
+    if (u <= 127) {
+      if (outIdx >= endIdx) break;
+      heap[outIdx++] = u;
+    } else if (u <= 2047) {
+      if (outIdx + 1 >= endIdx) break;
+      heap[outIdx++] = 192 | (u >> 6);
+      heap[outIdx++] = 128 | (u & 63);
+    } else if (u <= 65535) {
+      if (outIdx + 2 >= endIdx) break;
+      heap[outIdx++] = 224 | (u >> 12);
+      heap[outIdx++] = 128 | ((u >> 6) & 63);
+      heap[outIdx++] = 128 | (u & 63);
+    } else {
+      if (outIdx + 3 >= endIdx) break;
+      heap[outIdx++] = 240 | (u >> 18);
+      heap[outIdx++] = 128 | ((u >> 12) & 63);
+      heap[outIdx++] = 128 | ((u >> 6) & 63);
+      heap[outIdx++] = 128 | (u & 63);
+    }
+  }
+  heap[outIdx] = 0;
+  return outIdx - startIdx;
+}
+
+const UTF8Decoder = typeof TextDecoder !== "undefined" ? new TextDecoder("utf8") : undefined;
```

Comment:
I think we can depend on `TextDecoder` being present and don't need this safety, I think every browser that supports WASM has it. I guess then we can delete a bunch of code here? I believe there's also a `TextEncoder`?

Reply:
Yes, we definitely can. Then again, that part of the code is mostly glue-code that's copied straight from emscripten. I'm hoping that someday Pyodide will support that out of the box and then we'll be able to get rid of it.